src/Test/Expectation.elm

```diff
@@ -1,16 +1,41 @@
-module Test.Expectation exposing (Expectation(..), withGiven)
+module Test.Expectation exposing (Expectation(..), Reason(..), withGiven)
 
 
 type Expectation
     = Pass
-    | Fail String String
+    | Fail { given : String, description : String, reason : Reason }
+
+
+
+-- TODO: given : Maybe String
+
+
+type Reason
+    = Custom
+    | Equals String String
+    | Comparison String String
+      -- Expected, actual, (index of problem, expected element, actual element)
+    | ListDiff String String ( Int, String, String )
```

Comment:
The first two arguments are the entire lists. The tuple in the first differing element and its index. (Not sure what happens if one list is a prefix of the other...)